dbfopen.c

```diff
@@ -799,25 +830,67 @@ int SHPAPI_CALL DBFAddNativeFieldType(DBFHandle psDBF, const char *pszFieldName,
 
     const int nOldRecordLength = psDBF->nRecordLength;
     const int nOldHeaderLength = psDBF->nHeaderLength;
+    int *psDBF_realloc_ptr = NULL;
+    char *psDBF_char_realloc_ptr = NULL;
 
     /* -------------------------------------------------------------------- */
     /*      realloc all the arrays larger to hold the additional field      */
     /*      information.                                                    */
     /* -------------------------------------------------------------------- */
     psDBF->nFields++;
 
-    psDBF->panFieldOffset = STATIC_CAST(
+    psDBF_realloc_ptr = STATIC_CAST(
         int *, realloc(psDBF->panFieldOffset, sizeof(int) * psDBF->nFields));
 
-    psDBF->panFieldSize = STATIC_CAST(
+    if (!psDBF_realloc_ptr)
+    {
+        DBFHandleFree(psDBF);
```

Comment:
The call to DBFHandleFree() doesn't look right to me. That should rather be (and same in below instances)
```suggestion
        psDBF->nFields--;
```

Reply:
Thanks for the review @rouault. But don't we be losing access to earlier memory pointed by `psDBF->panFieldOffset` if we are resetting it in case realloc fails and therefore not being able to free it causing a memory leak?

Reply:
"If realloc() fails, the original block is left untouched; it is not freed or moved" , so as in that function we try to add a new field and grow related arrays, if we can't, just resetting nFields to the initial value should leave the library in a stable state

Reply:
Thank you. I see your point and it makes sense, I will make the changes.